### PR TITLE
EXSWHTEC-109 - Implement tests for the hipModuleLoad family of APIs and hipModuleUnload

### PR DESCRIPTION
**Pull request by music-dino**
- Implement positive and negative tests for hipModuleLoad
- Implement positive and negative tests for hipModuleLoadData
- Implement positive and negative tests for hipModuleLoadDataEx
- Implement positive tests for hipModuleUnload as part of the above tests
- Implement negative tests for hipModuleUnload.

### CODE REVIEW

**Review by chrispaquot — APPROVED** [review]

File: catch/unit/module/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    hipModuleLoadData.cc 
+    hipModuleLoadDataEx.cc 
+    hipModuleUnload.cc 
+)
+
+configure_file(not_a_module.txt not_a_module.txt)
+
+add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/empty_module.code
+                   COMMAND ${CMAKE_CXX_COMPILER} --genco --std=c++17 ${CMAKE_CURRENT_SOURCE_DIR}/empty_module.cc -o empty_module.code
+                   DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/empty_module.cc)
```
Comment (mangupta) [inline catch/unit/module/CMakeLists.txt]: genco does not work properly on windows yet. you will need to exclude this test building on windows. additionally you are missing specifying the offload arch & output path for building the code object. see unit/deviceLib/CMakeLists.txt for an example of how to do it for all of the above.

**Review by mangupta — CHANGES_REQUESTED** [review]

File: catch/unit/module/CMakeLists.txt (hunk 1 above)
Comment (music-dino) [reply catch/unit/module/CMakeLists.txt]: Requested changes have been implemented, please proceed.

File: catch/unit/module/hipModuleLoad.cc [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  SECTION("fname == empty string") {
+    HIP_CHECK_ERROR(hipModuleLoad(&module, ""), hipErrorInvalidValue);
+  }
+
+  SECTION("fname == non existent file") {
+    HIP_CHECK_ERROR(hipModuleLoad(&module, "non existent file"), hipErrorFileNotFound);
+  }
+
+// Disabled for AMD due to defect - EXSWHTEC-151
+#if HT_NVIDIA
```
Comment (chrispaquot) [inline catch/unit/module/hipModuleLoad.cc]: Create a new TEST_CAST for this and disable it via json.

File: catch/unit/module/hipModuleLoadData.cc [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    HIP_CHECK_ERROR(hipModuleLoadData(nullptr, loaded_module.data()), hipErrorInvalidValue);
+    LoadModuleIntoBuffer("empty_module.code");
+  }
+
+  SECTION("image == nullptr") {
+    HIP_CHECK_ERROR(hipModuleLoadData(&module, nullptr), hipErrorInvalidValue);
+  }
+
+// Disabled for AMD due to defect - EXSWHTEC-153
+#if HT_NVIDIA
```
Comment (chrispaquot) [inline catch/unit/module/hipModuleLoadData.cc]: Same here.

File: catch/unit/module/hipModuleLoadDataEx.cc [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    LoadModuleIntoBuffer("empty_module.code");
+  }
+
+  SECTION("image == nullptr") {
+    HIP_CHECK_ERROR(hipModuleLoadDataEx(&module, nullptr, 0, nullptr, nullptr),
+                    hipErrorInvalidValue);
+  }
+
+// Disabled for AMD due to defect - EXSWHTEC-153
+#if HT_NVIDIA
```
Comment (chrispaquot) [inline catch/unit/module/hipModuleLoadDataEx.cc]: again

File: catch/unit/module/hipModuleUnload.cc [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+*/
+
+#include <hip_test_common.hh>
+#include <hip/hip_runtime_api.h>
+
+TEST_CASE("Unit_hipModuleUnload_Negative_Parameters") {
+  HIP_CHECK(hipFree(nullptr));
+
+// Disabled for AMD due to defect - EXSWHTEC-152
+#if HT_NVIDIA
```
Comment (chrispaquot) [inline catch/unit/module/hipModuleUnload.cc]: again

File: catch/unit/module/not_a_module.txt
Comment (chrispaquot) [inline catch/unit/module/not_a_module.txt:1]: is the empty file necessary?

Comment (music-dino) [reply catch/unit/module/not_a_module.txt:1]: The test requires a file that exists in the filesystem, but does not represent a valid module file. This seemed like the most straightforward approach.

File: catch/unit/module/hipModuleLoad.cc (hunk 2 above)
Comment (music-dino) [reply catch/unit/module/hipModuleLoad.cc]: The requested change has been implemented.

File: catch/unit/module/hipModuleLoadData.cc (hunk 3 above)
Comment (music-dino) [reply catch/unit/module/hipModuleLoadData.cc]: The requested change has been implemented.

File: catch/unit/module/hipModuleLoadDataEx.cc (hunk 4 above)
Comment (music-dino) [reply catch/unit/module/hipModuleLoadDataEx.cc]: The requested change has been implemented.

File: catch/unit/module/hipModuleUnload.cc (hunk 5 above)
Comment (music-dino) [reply catch/unit/module/hipModuleUnload.cc]: The requested change has been implemented.

**Comment by rorake** [comment]
@music-dino 
Will you please resolve the merge conflicts for this PR?

**Comment by music-dino** [comment]
@rorake I don't see your comment for some reason, so I cannot quote reply to it. Merge conflicts have been resolved, please proceed.

**Comment by rakesroy** [comment]
PR has been merged into develop branch via commit 2c6d940ac05b4981acb4e9aa70c84d0de2443e0a.

**Pull request closed without merging**